File: anvil-api/src/main/java/org/anvilpowered/anvil/base/datastore/BaseCacheService.java
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package org.anvilpowered.anvil.base.datastore;

import org.anvilpowered.anvil.api.datastore.CacheService;
import org.anvilpowered.anvil.api.model.ObjectWithId;
import org.anvilpowered.anvil.api.registry.AnvilKeys;
import org.anvilpowered.anvil.api.registry.key.Keys;
import org.anvilpowered.anvil.api.registry.Registry;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.Set;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.ConcurrentMap;
import java.util.function.Predicate;
import java.util.stream.Collectors;

@Deprecated // will probably break in 0.4
public abstract class BaseCacheService<
    TKey,
    T extends ObjectWithId<TKey>,
    TDataStore>
    extends BaseRepository<TKey, T, TDataStore>
    implements CacheService<TKey, T, TDataStore> {

    protected Registry registry;

    protected ConcurrentMap<T, Long> cache;

    private Integer timeoutSeconds;

    protected BaseCacheService(Registry registry) {
        this.registry = registry;
//        registry.whenLoaded(this::registryLoaded).register();
        cache = new ConcurrentHashMap<>();
    }

    private void registryLoaded() {
        stopCacheInvalidationTask();
        Integer intervalSeconds = registry.get(AnvilKeys.INSTANCE.getCACHE_INVALIDATION_INTERVAL_SECONDS());
        timeoutSeconds = registry.get(AnvilKeys.INSTANCE.getCACHE_INVALIDATION_TIMOUT_SECONDS());
        startCacheInvalidationTask(intervalSeconds);
    }

    @Override
    public Runnable getCacheInvalidationTask() {
        return () -> {
            List<T> toRemove = new ArrayList<>();
            for (T t : getAllAsSet()) {
                if (System.currentTimeMillis() - cache.get(t) > timeoutSeconds * 1000L) {
                    toRemove.add(t);
                }
            }
            toRemove.forEach(this::delete);
        };
    }

    @Override
    public Set<T> getAllAsSet() {
        return cache.keySet();
    }

    @Override
    public CompletableFuture<Optional<T>> insertOne(T t) {
        return CompletableFuture.supplyAsync(() -> {
            if (t == null) return Optional.empty();
            cache.put(t, System.currentTimeMillis());
            return Optional.of(t);
        });
    }

    @Override
    public CompletableFuture<List<T>> insert(List<T> list) {
        return CompletableFuture.supplyAsync(() -> list.stream().map(t -> insertOne(t).join().orElse(null)).filter(Objects::nonNull).collect(Collectors.toList()));
    }

    @Override
    public CompletableFuture<Optional<T>> getOne(TKey id) {
        return CompletableFuture.supplyAsync(() -> getOne(dbo -> dbo.getId().equals(id)));
    }

    @Override
    public CompletableFuture<Boolean> deleteOne(TKey id) {
        return CompletableFuture.supplyAsync(() -> deleteOne(dbo -> dbo.getId().equals(id)).isPresent());
    }

    @Override
    public CompletableFuture<List<TKey>> getAllIds() {
        return CompletableFuture.supplyAsync(() -> cache.keySet().stream().map(ObjectWithId::getId).collect(Collectors.toList()));
    }

    @Override
    public Optional<T> deleteOne(Predicate<? super T> predicate) {
        return getOne(predicate).flatMap(this::delete);
    }

    @Override
    public Optional<T> delete(T t) {
        return cache.remove(t) == null ? Optional.empty() : Optional.of(t);
    }

    @Override
    public List<T> delete(Predicate<? super T> predicate) {
        return cache.keySet().stream().filter(predicate).filter(t -> cache.remove(t) != null).collect(Collectors.toList());
    }

    @Override
    public List<T> getAll(Predicate<? super T> predicate) {
        return getAllAsSet().stream().filter(predicate).collect(Collectors.toList());
    }

    @Override
    public Optional<T> getOne(Predicate<? super T> predicate) {
        return getAll(predicate).stream().findAny();
    }
}

File: api/src/main/kotlin/org/anvilpowered/anvil/base/datastore/BaseCacheService.kt
/*
 *   Anvil - AnvilPowered
 *   Copyright (C) 2020-2021
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Lesser General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU Lesser General Public License for more details.
 *
 *     You should have received a copy of the GNU Lesser General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package org.anvilpowered.anvil.base.datastore

import org.anvilpowered.anvil.api.registry.Keys
import org.anvilpowered.anvil.api.registry.Registry
import java.util.Objects
import java.util.Optional
import java.util.function.Consumer
import java.util.function.Function
import java.util.function.Predicate
import java.util.function.Supplier

@Deprecated("") // will probably break in 0.4
abstract class BaseCacheService<TKey, T : ObjectWithId<TKey>?, TDataStore> protected constructor(protected var registry: Registry) :
    BaseRepository<TKey, T, TDataStore>(), CacheService<TKey, T, TDataStore> {
    protected var cache: ConcurrentMap<T, Long>
    private var timeoutSeconds: Int? = null

    init {
        registry.whenLoaded { registryLoaded() }.register()
        cache = ConcurrentHashMap<T, Long>()
    }

    private fun registryLoaded() {
        stopCacheInvalidationTask()
        val intervalSeconds = registry.getOrDefault<Int>(Keys.CACHE_INVALIDATION_INTERVAL_SECONDS)
        timeoutSeconds = registry.getOrDefault<Int>(Keys.CACHE_INVALIDATION_TIMOUT_SECONDS)
        startCacheInvalidationTask(intervalSeconds)
    }

    val cacheInvalidationTask: Runnable
        get() = Runnable {
            val toRemove: MutableList<T> = ArrayList()
            for (t in allAsSet) {
                if (System.currentTimeMillis() - cache.get(t) > timeoutSeconds!! * 1000L) {
                    toRemove.add(t)
                }
            }
            toRemove.forEach(Consumer { t: T -> this.delete(t) })
        }
    val allAsSet: Set<T>
        get() = cache.keys

    override fun insertOne(t: T): CompletableFuture<Optional<T>?>? {
        return CompletableFuture.supplyAsync<Optional<T>>(Supplier {
            if (t == null) return@supplyAsync Optional.empty<T>()
            cache.put(t, System.currentTimeMillis())
            Optional.of<T>(t)
        })
    }

    override fun insert(list: List<T>?): CompletableFuture<List<T>?>? {
        return CompletableFuture.supplyAsync<List<T>>(Supplier {
            list!!.stream().map(
                Function<T, T> { t: T -> insertOne(t).join().orElse(null) }).filter { obj: T -> Objects.nonNull(obj) }
                .collect(Collectors.toList<T>())
        })
    }

    override fun getOne(id: TKey): CompletableFuture<Optional<T>> {
        return CompletableFuture.supplyAsync<Optional<T>>(Supplier { getOne { dbo: T -> dbo.getId() == id } })
    }

    override fun deleteOne(id: TKey): CompletableFuture<Boolean?>? {
        return CompletableFuture.supplyAsync<Boolean>(Supplier { deleteOne { dbo: T -> dbo.getId() == id }!!.isPresent })
    }

    override val allIds: CompletableFuture<List<TKey>>
        get() = CompletableFuture.supplyAsync<List<TKey>>(Supplier<List<TKey>> {
            cache.keys.stream().map<TKey>(
                Function<T, TKey> { obj: T -> obj.getId() }).collect<List<TKey>, Any>(Collectors.toList<TKey>())
        })

    override fun deleteOne(predicate: Predicate<in T>?): Optional<T>? {
        return getOne(predicate)!!.flatMap { t: T -> this.delete(t) }
    }

    override fun delete(t: T): Optional<T>? {
        return if (cache.remove(t) == null) Optional.empty() else Optional.of(t)
    }

    override fun delete(predicate: Predicate<in T>?): List<T>? {
        return cache.keys.stream().filter(predicate).filter(Predicate { t: T -> cache.remove(t) != null })
            .collect<List<T>, Any>(Collectors.toList<T>())
    }

    override fun getAll(predicate: Predicate<in T>?): List<T>? {
        return allAsSet.stream().filter(predicate).collect(Collectors.toList<T>())
    }

    override fun getOne(predicate: Predicate<in T>?): Optional<T>? {
        return getAll(predicate)!!.stream().findAny()
    }
}
